### PR TITLE
Integrate torchax custom attention kernel into ulysses

### PR DESCRIPTION
**Pull request by eltsai**
Adding [torchax path's custom kernel](https://github.com/Perseus14/diffusers-exp/blob/wan2.2-main/exp/custom_splash_attention_updated.py) into ulysses (triggered when attention=`ulysses_custom`)

Inference time:
```
==================================================
  TIMING SUMMARY
==================================================
  Load (checkpoint):      80.6s
  Compile:               186.9s
  ────────────────────────────────────────
  Inference:             167.2s
==================================================
```

### CODE REVIEW

File: src/maxdiffusion/models/attention_flax.py [hunk 1]
```diff
@@ -641,6 +642,94 @@ def wrap_ulysses_attention(query, key, value):
...
+  def wrap_ulysses_attention(query, key, value):
+    query = jax.lax.all_to_all(query, axis_name=axis_name, split_axis=1, concat_axis=2, tiled=True)
+    key = jax.lax.all_to_all(key, axis_name=axis_name, split_axis=1, concat_axis=2, tiled=True)
+    value = jax.lax.all_to_all(value, axis_name=axis_name, split_axis=1, concat_axis=2, tiled=True)
+
+    bq = 2048
+    bkv = 2048
+    bkv_compute = 1024
+    bkv_compute_in = 256
+    heads_per_tile = 1
```
Comment (Perseus14) [inline src/maxdiffusion/models/attention_flax.py]: Updating this to

    bq = 4864
    bkv = 1024
    bkv_compute = 1024
    bkv_compute_in = 1024
    heads_per_tile = 1

and using this [command](https://paste.googleplex.com/6567116141625344) gave me the following latency

```
  Load (checkpoint):     297.0s
  Compile:               219.8s
  ───────────────────────────────
  Inference:             147.4s
```

File: src/maxdiffusion/kernels/custom_splash_attention.py
Comment (Perseus14) [inline src/maxdiffusion/kernels/custom_splash_attention.py:1]: Please note that this custom kernel does not support `attenton_masks`

File: src/maxdiffusion/kernels/custom_splash_attention.py [hunk 2]
```diff
@@ -0,0 +1,682 @@
+"""Custom Pallas flash attention kernel for TPU."""
```
Comment (entrpn) [inline src/maxdiffusion/kernels/custom_splash_attention.py:17]: this file needs a license header.

File: src/maxdiffusion/kernels/custom_splash_attention.py [hunk 3]
```diff
@@ -0,0 +1,682 @@
...
+from jax.experimental.shard_map import shard_map
+from jax.sharding import PartitionSpec as P
+
+DEFAULT_MASK_VALUE = -0.7 * float(np.finfo(np.dtype("float32")).max)
+NUM_LANES = 128
+NUM_SUBLANES = 8
+NT_DIM_NUMBERS = (((1,), (1,)), ((), ()))
+
+# Default block sizes (tuned for 720p Wan2.1 on v6e/v7x)
+DEFAULT_BQSIZE = 3328
```
Comment (entrpn) [inline src/maxdiffusion/kernels/custom_splash_attention.py:37]: based on @Perseus14 comments about changing block sizes, do these values also need to be updated here?

File: src/maxdiffusion/models/attention_flax.py [hunk 4]
```diff
@@ -641,6 +642,94 @@ def wrap_ulysses_attention(query, key, value):
...
+  kv_axis_names = nn.logical_to_mesh_axes(axis_names_kv)
+
+  @functools.partial(
+      jax.shard_map,
+      mesh=mesh,
+      in_specs=(q_axis_names, kv_axis_names, kv_axis_names),
+      out_specs=q_axis_names,
+      check_vma=False,
+  )
+  def wrap_ulysses_attention(query, key, value):
```
Comment (entrpn) [inline src/maxdiffusion/models/attention_flax.py]: is this function being replicated? If so, is it possible to consolidate it with ulysses attention.

**Comment by eltsai** [comment]
Updated stats:
| Accelerator | Sharding | E2E time | log | Video |
| -- | -- | -- | -- | -- |
| v7x-8| dp2-context4-tp1| 139.3s | [log](https://paste.googleplex.com/6304394032316416) | [Video](https://drive.google.com/file/d/1S3gOFND6jbpSp1uSt_XOjyWm0FyCew97/view?usp=drive_link) |
| v7x-16| dp2-context8-tp1| 70.2s | [log](https://pantheon.corp.google.com/logs/query;query=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22cloud-tpu-inference-test%22%0Aresource.labels.location%3D%22us-central1%22%0Aresource.labels.cluster_name%3D%22elisatsai-vm%22%0Aresource.labels.namespace_name%3D%22default%22%0Aresource.labels.pod_name:%22wan22-ulysses-custom-cusk-new-flags2-slice-job-0-0-%22%0Aseverity%3E%3DDEFAULT;storageScope=project;cursorTimestamp=2026-04-27T18:58:03.431077035Z;duration=P1D?project=cloud-tpu-inference-test&e=13802955&mods=allow_workbench_image_override) | [Video](https://drive.google.com/file/d/1gKeF2Bjm85V99pjOv6TA_OyXgglj8Q6X/view?usp=drive_link) |

**Review by entrpn — DISMISSED** [review]

File: src/maxdiffusion/models/custom_splash_attention.py [hunk 5]
```diff
@@ -0,0 +1,698 @@
...
+      ),
+      grid_spec=pltpu.PrefetchScalarGridSpec(
+          num_scalar_prefetch=0,
+          in_specs=in_specs,
+          out_specs=out_specs,
+          grid=grid,
+      ),
+      compiler_params=pltpu.CompilerParams(
+          dimension_semantics=("parallel", "arbitrary", "arbitrary"),
+          flags={"XLA_TPU_FORCE_LP_LLO_SCHEDULER": True},
```
Comment (Perseus14) [inline src/maxdiffusion/models/custom_splash_attention.py]: I believe tokamax's `use_experimental_scheduler` enables this XLA flag. Could we use the config to set this param?

File: src/maxdiffusion/models/custom_splash_attention.py [hunk 6]
```diff
@@ -0,0 +1,698 @@
...
+# Default block sizes (tuned for 720p Wan2.1 on v6e/v7x)
+DEFAULT_BQSIZE = 3328
+DEFAULT_BKVSIZE = 2816
+# Cranked up to 1024 for massive MXU throughput
+DEFAULT_BKVCOMPUTESIZE = 1024
+# Kept at 256 to protect VPU registers (V1 Optimization)
+DEFAULT_BKVCOMPUTEINSIZE = 256
+
+
+def _exp2(x: jax.Array) -> jax.Array:
```
Comment (Perseus14) [inline src/maxdiffusion/models/custom_splash_attention.py]: Similar to `tokamax` we can set config `use_base2_exp` to toggle between `jnp.exp` and `jnp.exp2`

```
LOG2E = math.log2(math.e)
LOG2E_INV = 1 / LOG2E

....

if config.use_base2_exp
    query = query * LOG2E

....

exp = jnp.exp2 if config.use_base2_exp else jnp.exp

```

File: src/maxdiffusion/models/attention_flax.py [hunk 7]
```diff
@@ -563,65 +567,93 @@ def wrap_ulysses_attention(query, key, value):
...
+      bkv_compute = 1024
+      bkv_compute_in = 1024
+      heads_per_tile = 1
 
-    block_q = max(*block_q_sizes)
-    query, kv_size, query_seq_len = _pad_data_for_flash(query, heads, block_q)
-    block_kv = max(*block_kv_sizes)
-    key, _, key_seq_len = _pad_data_for_flash(key, heads, block_kv)
-    value, _, _ = _pad_data_for_flash(value, heads, block_kv)
+      query_scaled = query * 1.44269504
```
Comment (Perseus14) [inline src/maxdiffusion/models/attention_flax.py]: Can we scale it inside the custom kernel? This scale is only applicable if use `base2_exp`

File: src/maxdiffusion/models/attention_flax.py [hunk 8]
```diff
@@ -563,65 +567,93 @@ def wrap_ulysses_attention(query, key, value):
...
-      block_kv_sizes += (block_sizes.block_kv_dkv,)
-    else:
-      block_q_sizes += (block_sizes.block_q_dq,)
-      block_kv_sizes += (block_sizes.block_kv_dq,)
+    if use_custom_kernel:
+      bq = 4864
+      bkv = 1024
+      bkv_compute = 1024
+      bkv_compute_in = 1024
+      heads_per_tile = 1
```
Comment (Perseus14) [inline src/maxdiffusion/models/attention_flax.py:577]: Is it possible to modify `flash_block_sizes` config so that users can set these values?

**Review by Perseus14 — COMMENTED** [review]
@eltsai I have left a few comments on the code. I am hoping we can reuse the user defined configs like `use_base2_exp`, `use_experimental_scheduler` and `flash_block_sizes` for the custom kernel.

Let me know if we can incorporate these changes now or attempt them later

cc: @entrpn

File: src/maxdiffusion/models/custom_splash_attention.py (hunk 5 above)
Comment (eltsai) [reply src/maxdiffusion/models/custom_splash_attention.py]: sure, I think this make sense. I will send out another PR

**Comment by eltsai** [comment]
Updated the PR based on @Perseus14 's comments: passing `use_base2_exp`, `use_experimental_scheduler` and `flash_block_sizes` instead of hard coded.

Also moved the custom kernel file under `src/maxdiffusion/kernels`

Updated the profiling code to log perf with more granularity:
```
==================================================
  TIMING SUMMARY
==================================================
  Load (checkpoint):     130.9s
  Compile:               137.5s
  ────────────────────────────────────────
  Inference:             137.4s
  Conditioning:            2.4s
  Denoise Total:         121.8s
  VAE Decode:             13.3s
==================================================
```

**Comment by eltsai** [comment]
Updated perf logging code. It was not logging each component's time correctly. In `WanPipeline2_2.__call__`, we start the denoise_total timer, run the loop, and stop it. Because of jax's asynchronous execution, the timer stops before the TPU has actually finished executing all 40 steps.

New run:
```
==================================================
  TIMING SUMMARY
==================================================
  Load (checkpoint):     128.6s
  Compile:               137.6s
  ────────────────────────────────────────
  Inference:             136.9s
  Conditioning:            2.2s
  Denoise Total:         130.6s
  VAE Decode:              4.0s
==================================================
```

Also added the two new flags from @rishabhmanoj
```
disable_bounds_checks=True,
skip_device_barrier=True,
```
which give us about ~1 sec of gain

**Review by entrpn — APPROVED** [review]
